### ISSUE TITLE
Freshly generated instance by reflection cannot be searched

### ISSUE THREAD

**Issue by L-TChen** [agda/agda#5757]
If a name is generated by reflection using `freshName`, then it cannot be used as an instance even though it is declared so. For example, consider the following record type `_HasNat` which merely contains a natural number:
```agda
open import Agda.Builtin.Reflection

open import Reflection
open import Data.Unit
open import Data.Nat
open import Data.List
open Clause
Tactic = Term → TC ⊤

record _HasNat (d : Name) : Set where
  constructor has
  field
    a : ℕ

find : ∀ d ⦃ _ : d HasNat ⦄ → ℕ
find d ⦃ has a ⦄ = a
```

Then, for every `Name` an instance is defined by macro with its name freshly generated:
```agda
pattern `has x     = con (quote has)     (vArg x ∷ [])
pattern _`HasNat x = def (quote _HasNat) (vArg x ∷ [])
pattern `42        = lit (nat 42)

derive : (d : Name) → TC Name
derive d = do
  `d   ← quoteTC d
  inst ← freshName "HasNat"
  declareDef (iArg inst) (`d `HasNat)
  defineFun inst [ clause [] [] (`has `42) ]
  return inst

macro
  deriveNat : Name → Tactic
  deriveNat d hole = do
    i ← derive d 
    unify hole (def i [])

⊤hasNat = deriveNat ⊤

T : ℕ
T = find (quote ⊤) -- doesn't work! 
```

As far as I know, I understand that the name generated by macro cannot be used directly since scope checking happens before type checking. However, in this case, instance resolution also happen during type checking, but it seems that the generated name does not become a candidate. 

Is this behaviour intended? I think it is reasonable to expect the generated instance accessible during instance resolution...

**Comment by jespercockx** [comment]
Instance search requires that the name of the instance is in scope (unqualified or qualified) for it to be considered. This prevents e.g. private instances from other modules from polluting the instance scope. However, it seems likely that names generated by `freshName` are also not considered to be in scope (since the user cannot refer to them) and hence the same logic in the instance search algorithm prevents them from being considered.

**Comment by L-TChen** [comment]
If the instance search algorithm considers names generated by `freshName`, then it is possible to define Haskell-style `deriving` as a macro. I am testing different usages of my library in development, and it seems that `deriving` could be useful.

Could you point out where to start if I’d like to change this behaviour? Adding an option at least should be okay?

**Comment by jespercockx** [comment]
The currently supported way of defining `deriving`-like macros in Agda is by using `unquoteDecl` or `unquoteDef`. What is the problem with using those instead (except obviously the fact that you are required to give a name to the instance yourself)?

**Comment by L-TChen** [comment]
Yes, I can use `unquoteDecl` to define `deriving`-like macros, but I don’t really want to give a bunch of names which can be inferred and are never referred explicitly.

**Comment by jespercockx** [comment]
Fair enough, though there's good reasons for wanting each definition to have an actual name that's written in the source code, e.g. for showing them in error messages.

If you anyways would like to experiment with changing instance search so it also considers names generated by `freshName`, the point to start would be at https://github.com/agda/agda/blob/master/src/full/Agda/TypeChecking/InstanceArguments.hs#L152
Note that just removing the check would also allow `private` instances from other modules, so it should be replaced with a different check instead.
